JOSESwift/Sources/AESCBCEncryption.swift

```diff
@@ -26,29 +26,48 @@ import Foundation
 struct AESCBCEncryption {
     private let contentEncryptionAlgorithm: ContentEncryptionAlgorithm
     private let contentEncryptionKey: Data
+    private let hmacKey: Data
+    private let hmacAlgorithm: HMACAlgorithm
 
-    init(contentEncryptionAlgorithm: ContentEncryptionAlgorithm, contentEncryptionKey: Data) {
+    init(contentEncryptionAlgorithm: ContentEncryptionAlgorithm, contentEncryptionKey: Data) throws {
+        guard contentEncryptionAlgorithm.checkKeyLength(for: contentEncryptionKey) else {
+            throw JWEError.keyLengthNotSatisfied
+        }
         self.contentEncryptionAlgorithm = contentEncryptionAlgorithm
-        self.contentEncryptionKey = contentEncryptionKey
+        // the key comes in as a combined key, split it according to the algorithm
+        switch contentEncryptionAlgorithm {
+        case .A256CBCHS512:
+            self.hmacKey = contentEncryptionKey.subdata(in: 0..<32)
+            self.contentEncryptionKey = contentEncryptionKey.subdata(in: 32..<64)
+        case .A128CBCHS256:
+            self.hmacKey = contentEncryptionKey.subdata(in: 0..<16)
+            self.contentEncryptionKey = contentEncryptionKey.subdata(in: 16..<32)
+        case .A256GCM, .A128GCM:
+            throw JWEError.contentEncryptionAlgorithmMismatch
+        }
+        // now select the appropriate hashing algorithm
+        switch contentEncryptionAlgorithm {
+        case .A256CBCHS512:
+            hmacAlgorithm = .SHA512
+        case .A128CBCHS256:
+            hmacAlgorithm = .SHA256
+        case .A256GCM, .A128GCM:
+            throw JWEError.contentEncryptionAlgorithmMismatch
+        }
```

Comment:
Totally a taste thing, not a change request:

Should we move these two switches to separate functions just like the `getAuthenticationTag` below? For me personally I feel like it's easier to keep an overview that way. But Both options work for me. Again, it's a taste thing.

Reply:
externalised both of them to functions. I like it better as well.